coderd/externalauth_test.go

```diff
@@ -706,4 +706,81 @@ func TestExternalAuthCallback(t *testing.T) {
 		})
 		require.NoError(t, err)
 	})
+	t.Run("AgentAPIKeyScope", func(t *testing.T) {
+		t.Parallel()
+
+		for _, tt := range []struct {
+			apiKeyScope  string
+			expectsError bool
+		}{
+			{apiKeyScope: "default", expectsError: false},
+			{apiKeyScope: "no_user_data", expectsError: true},
+		} {
+			t.Run(tt.apiKeyScope, func(t *testing.T) {
+				t.Parallel()
+
+				client := coderdtest.New(t, &coderdtest.Options{
+					IncludeProvisionerDaemon: true,
+					ExternalAuthConfigs: []*externalauth.Config{{
+						InstrumentedOAuth2Config: &testutil.OAuth2Config{},
+						ID:                       "github",
+						Regex:                    regexp.MustCompile(`github\.com`),
+						Type:                     codersdk.EnhancedExternalAuthProviderGitHub.String(),
+					}},
+				})
+				user := coderdtest.CreateFirstUser(t, client)
+				authToken := uuid.NewString()
+				version := coderdtest.CreateTemplateVersion(t, client, user.OrganizationID, &echo.Responses{
+					Parse:          echo.ParseComplete,
+					ProvisionPlan:  echo.PlanComplete,
+					ProvisionApply: echo.ProvisionApplyWithAgentAndAPIKeyScope(authToken, tt.apiKeyScope),
+				})
+				template := coderdtest.CreateTemplate(t, client, user.OrganizationID, version.ID)
+				coderdtest.AwaitTemplateVersionJobCompleted(t, client, version.ID)
+				workspace := coderdtest.CreateWorkspace(t, client, template.ID)
+				coderdtest.AwaitWorkspaceBuildJobCompleted(t, client, workspace.LatestBuild.ID)
+
+				agentClient := agentsdk.New(client.URL)
+				agentClient.SetSessionToken(authToken)
+
+				token, err := agentClient.ExternalAuth(t.Context(), agentsdk.ExternalAuthRequest{
+					Match: "github.com/asd/asd",
+				})
+				require.NoError(t, err)
+				require.NotEmpty(t, token.URL)
```

Comment:
This should fail quicker.

The reason it's pending at all, is because if a link does not exist, then we get a 404, instead of an RBAC error. And a 404 is an expected case:

https://github.com/coder/coder/blob/2bdc47066248b35b48916d43b0b62bb0b400d890/coderd/workspaceagents.go#L1609-L1626

I made a PR into this branch that rejects early for forbidden: https://github.com/coder/coder/pull/17712

If you merge this in, update this test to expect the error up front.

Reply:
Can you also assert the error received is a forbidden error?